server/controllers/finance/reports/income_expense_by_month/index.js

```diff
@@ -0,0 +1,304 @@
+      FROM period_total AS pt
+      JOIN period AS p ON p.id = pt.period_id
+      WHERE pt.fiscal_year_id = ?
+        AND p.number BETWEEN ? AND ?
+      GROUP BY pt.account_id
+    )s ON a.id = s.account_id
+    WHERE a.type_id IN (?)
+    ORDER BY a.number;
+  `;
+
+  const accountTypes = [
+    TITLE_ID,
+    INCOME_TYPE_ID,
+    EXPENSE_TYPE_ID,
+  ];
+
+  return db.exec(sql, [fiscalYearId, periodFromNumber, periodToNumber, accountTypes]);
+}
+
+/**
+ * @function getPeriodById
+ *
+ * @description
+ * Small helper to get all properties of a period by its id.
+ */
+function getPeriodById(id) {
+  const sql = `
+    SELECT id, number, start_date, end_date, locked FROM period WHERE id = ?;
+  `;
+
+  return db.one(sql, id);
+}
+
+/**
+ * @function combineIntoSingleDataset
+ *
+ * @description
+ * This function takes two partially overlapping datasets and combines them into
+ * a single dataset.  The current balances set the balance values and the
+ * previous balances are assigned to previousBalance property.  Missing data is
+ * filled in with zeroes.
+ *
+ * @returns Array
+ */
+function combineIntoSingleDataset(currentBalances, previousBalances, nextBalances) {
+  // make an id -> account map for both values
+  const currentMap = new Map(currentBalances.map(a => [a.id, a]));
+  const previousMap = new Map(previousBalances.map(a => [a.id, a]));
+  const nextMap = new Map(nextBalances.map(a => [a.id, a]));
+
+  // get all account ids in a single list.
+  const combined = _.uniq([...currentMap.keys(), ...previousMap.keys(), ...nextMap.keys()]);
+
+  return combined.map(id => {
+    const record = {};
+
+    const current = currentMap.get(id);
+    const previous = previousMap.get(id);
+    const next = nextMap.get(id);
+
+    if (current) {
+      const previousBalance = previous ? previous.balance : 0;
+      const nextBalance = next ? next.balance : 0;
+      _.extend(record, current, { previousBalance, nextBalance });
+    } else if (previous) {
+      const nextBalance = next ? next.balance : 0;
+      _.extend(record, previous, { balance : 0 }, { previousBalance : previous.balance, nextBalance });
+    } else {
+      _.extend(record, next, { balance : 0, previousBalance : 0 }, { nextBalance : next.balance });
+    }
+
+
+    return record;
+  });
+}
+
+/**
+ * @function getPeriodByNumberAndFiscalId
+ *
+ * @description
+ * Small helper to get all properties of a period by its number.
+ */
+function getPeriodByNumberAndFiscalId(number, fiscalId) {
+  const sql = `
+    SELECT id, number, start_date, end_date, locked FROM period WHERE number = ? AND fiscal_year_id = ?;
+  `;
+
+  return db.one(sql, [number, fiscalId]);
+}
+
+/**
+ * @function constructAndPruneTree
+ *
+ * @description
+ * Receives a dataset of balances and creates the account tree.
+ */
+function constructAndPruneTree(dataset) {
+  const tree = new Tree(dataset);
+
+  const properties = [
+    'balance', 'previousBalance', 'difference', 'variance',
```

Comment:
You will need to add `nextBalance` to this list to compute totals on that column.